src/ecmult_impl.h

```diff
@@ -56,44 +56,51 @@
 
 #define ECMULT_MAX_POINTS_PER_BATCH 5000000
 
-/** Fill a table 'prej' with precomputed odd multiples of a. Prej will contain
- *  the values [1*a,3*a,...,(2*n-1)*a], so it space for n values. zr[0] will
- *  contain prej[0].z / a.z. The other zr[i] values = prej[i].z / prej[i-1].z.
- *  Prej's Z values are undefined, except for the last value.
+/** Fill a table 'pre_a' with precomputed odd multiples of a.
+ *  pre_a will contain [1*a,3*a,...,(2*n-1)*a], so it needs space for n group elements.
+ *  zr needs space for n field elements.
+ *
+ *  Although pre_a is an array of _ge rather than _gej, it actually represents elements
+ *  in Jacobian coordinates with their z coordinates omitted. The omitted z-coordinates
+ *  can be recovered using z and zr. Using the notation z(b) to represent the omitted
+ *  z coordinate of b:
+ *  - z(pre_a[n-1]) = 'z'
+ *  - z(pre_a[i-1]) = z(pre_a[i]) / zr[i] for n > i > 0
+ *
+ *  Lastly the zr[0] value, which isn't used above, is set so that:
+ *  - a.z = z(pre_a[0]) / zr[0]
  */
-static void secp256k1_ecmult_odd_multiples_table(int n, secp256k1_gej *prej, secp256k1_fe *zr, const secp256k1_gej *a) {
-    secp256k1_gej d;
-    secp256k1_ge a_ge, d_ge;
+static void secp256k1_ecmult_odd_multiples_table(int n, secp256k1_ge *pre_a, secp256k1_fe *zr, secp256k1_fe *z, const secp256k1_gej *a) {
+    secp256k1_gej d, ai;
+    secp256k1_ge d_ge;
     int i;
 
     VERIFY_CHECK(!a->infinity);
 
     secp256k1_gej_double_var(&d, a, NULL);
 
     /*
-     * Perform the additions on an isomorphism where 'd' is affine: drop the z coordinate
-     * of 'd', and scale the 1P starting value's x/y coordinates without changing its z.
+     * Perform the additions using an isomorphism that divides the z coordinate by the
+     * constant d.z. The group law is the same in the image, and since 'd' maps to an
+     * affine point, addition becomes more efficient (using mixed coordinates).
```

Comment:
I'm having trouble following your suggested wording.

We are working on the isomorphic curve y^2 = x^3 + 7*C^6 where C=d.z with our isomorphism sending secp256k1 to this curve is defined by phi(x,y) = (x*C^2, y*C^3).  The transformation of some generic point in Jacobian coordinates a := (ax, ay, az) is phi(a) = (ax*C^2,ay*C^3,az) or equivalently phi(a) = (ax,ay,az/C).  In particular a set of Jacobian coordinates for phi(d) is (x,y,1) which means the affine coordinates of phi(d) is (x,y).  The inverse isomorphism can be applied to a point in Jacobian coordinates by multiplying the z coordinate by C.

We have algorithms for adding two generic points in Jacobian coordinates, and a slightly optimized algorithm for adding a point in Jacobian coordinates with a point in affine coordinates.  The optimization amounts to noting that the z value of the point in affine coordinates is 1 and then removing all the places where you would otherwise multiply by 1.  Note that we are not adding two affine points, and thus field inversions are not under consideration, nor were they ever under consideration.

That all said, maybe I can try to expand upon this comment to make it more clear.  Doing operations on an isomorphic curve is definitely hard to follow.

Reply:
Sorry for the confusion. Your description of the isomorphism is correct. I am looking for a concise explanation of the optimization which is that we can use the faster `gej_add_ge` instead of `gej_add`. Surely we could use the former algorithm without the transformation by doing a field inversion to convert `d` into a regular `ge`. Perhaps this point is only confusing and not noteworthy.

Reply:
Your new explanation is very good.